### PR TITLE
CLOUDSTACK-3035: Implemented capabilities for snapshots and vm snapshots.

### PR DESCRIPTION
**Pull request by bwsw**
## Description

Implemented additional capabilities:

```
disksnapshotsmaxdaily = 6
disksnapshotsmaxhourly = 1
disksnapshotsmaxmonthly = 24
disksnapshotsmaxweekly = 12
vmsnapshotexpireinterval = 3600
vmsnapshotsmax = 1

```

Ticket: #3035 

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

As the feature requires MS to restart, tested manually.

### CODE REVIEW

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java [hunk 1]
```diff
@@ -1272,15 +1273,15 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
...
+        Type.HOURLY.setMax(snapshotHourlyMax.value());
+        Type.DAILY.setMax(snapshotDailyMax.value());
+        Type.WEEKLY.setMax(snapshotWeeklyMax.value());
+        Type.MONTHLY.setMax(snapshotMonthlyMax.value());
         _totalRetries = NumbersUtil.parseInt(_configDao.getValue("total.retries"), 4);
         _pauseInterval = 2 * NumbersUtil.parseInt(_configDao.getValue("ping.interval"), 60);
 
-        snapshotBackupRetries = BackupRetryAttempts.value();
-        snapshotBackupRetryInterval = BackupRetryInterval.value();
+        snapshotBackupRetries = backupRetryAttempts.value();
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: 'retries' would be 'attempts' minus 1, would it?

File: engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java [hunk 2]
```diff
@@ -28,14 +28,35 @@
...
 
-    static final ConfigKey<Integer> VMSnapshotExpireInterval = new ConfigKey<Integer>("Advanced", Integer.class, "vmsnapshot.expire.interval", "-1",
-            "VM Snapshot expire interval in hours", true, ConfigKey.Scope.Account);
+    Integer defaultVirtualMachineSnapshotsMax = 10;
+    Integer defaultVirtualMachineJobCheckInterval = 3000;
+    Integer defaultVirtualMachineSnapshotExpireInterval = -1;
+    Integer defaultVirtualMachineSnapshotCreateWait = 1800;
 
-    public static final int VMSNAPSHOTMAX = 10;
+    ConfigKey<Integer> virtualMachineSnapshotExpireInterval = new ConfigKey<Integer>("Advanced",
```
Comment (DaanHoogland) [inline engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java]: isn't convention for config-values capatalised?

File: engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java [hunk 3]
```diff
@@ -45,7 +66,7 @@
      * Sync VM snapshot state when VM snapshot in reverting or snapshoting or expunging state
      * Used for fullsync after agent connects
      *
-     * @param vm, the VM in question
+     * @param vm,    the VM in question
```
Comment (DaanHoogland) [inline engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java]: since you are fixing the java doc; pleae remove the comma.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java (hunk 1 above)
Comment (bwsw) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: @DaanHoogland 
1. #3035 thorough description added

Comment (bwsw) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: @DaanHoogland this is no my code, don't dive into semantics. What I did is tjust renamed variable to comply normal camel case style with refactoring utility.

If I didn't get what you mean, please try to explain me one more time)

File: engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java (hunk 2 above)
Comment (bwsw) [reply engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java]: @DaanHoogland Well, there are different kinds of coding conventions I have met in the codebase. Why I use the style like that here? The answer is because those definitions are used only once to define ConfigKey variables and some kind of syntax sugar which improves manageability.
If it should be changed to a capitalized style, can do that.

File: engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java (hunk 3 above)
Comment (bwsw) [reply engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java]: @DaanHoogland Done

File: engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java (hunk 2 above)
Comment (DaanHoogland) [reply engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java]: ok, not making a point of it.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java (hunk 1 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: ok, not your doing, @bwsw. rather ugly misuse of English though: number of attempts == initial attempt plus number of retries.

**Comment by bwsw** [comment]
@rhtyd @rafaelweingartner @DaanHoogland 
Just friendly reminder. Could you guys please to put milestone label, make a review, give comments?

File: api/src/main/java/org/apache/cloudstack/api/command/user/config/ListCapabilitiesCmd.java [hunk 4]
```diff
@@ -59,6 +59,15 @@ public void execute() {
         response.setKVMSnapshotEnabled((Boolean)capabilities.get("KVMSnapshotEnabled"));
         response.setAllowUserViewDestroyedVM((Boolean)capabilities.get("allowUserViewDestroyedVM"));
         response.setAllowUserExpungeRecoverVM((Boolean)capabilities.get("allowUserExpungeRecoverVM"));
+
+        response.setVmSnapshotsMax((Integer)capabilities.get("vmsnapshotsmax"));
```
Comment (rafaelweingartner) [inline api/src/main/java/org/apache/cloudstack/api/command/user/config/ListCapabilitiesCmd.java]: What about extracting these names to enum? and then documenting the option/names to better explain what they mean.

File: api/src/main/java/org/apache/cloudstack/api/response/CapabilitiesResponse.java [hunk 5]
```diff
@@ -84,6 +84,30 @@
     @Param(description = "true if the user can recover and expunge virtualmachines, false otherwise", since = "4.6.0")
     private boolean allowUserExpungeRecoverVM;
 
+    @SerializedName("vmsnapshotsmax")
```
Comment (rafaelweingartner) [inline api/src/main/java/org/apache/cloudstack/api/response/CapabilitiesResponse.java]: Same constant from the other class

File: engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java [hunk 6]
```diff
@@ -28,15 +28,36 @@
...
+            "Maximum time to wait for a VM snapshot creation in seconds",
+            false);
 
     /**
      * Delete all VM snapshots belonging to one VM
-     * @param id, VM id
-     * @param type,
+     *
+     * @param id   VM id
+     * @param type
```
Comment (rafaelweingartner) [inline engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java]: what does type mean?
Also, there is an excessive use of space in the "id" description

File: engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java [hunk 7]
```diff
@@ -45,7 +66,7 @@
      * Sync VM snapshot state when VM snapshot in reverting or snapshoting or expunging state
      * Used for fullsync after agent connects
      *
-     * @param vm, the VM in question
+     * @param vm    the VM in question
```
Comment (rafaelweingartner) [inline engine/components-api/src/main/java/com/cloud/vm/snapshot/VMSnapshotManager.java]: why are you adding these extra spaces?

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManager.java [hunk 8]
```diff
@@ -34,27 +34,84 @@
...
+    Integer defaultDiskSnapshotMaxHourly = 8;
+    Integer defaultDiskSnapshotMaxDaily = 8;
+    Integer defaultDiskSnapshotMaxWeekly = 8;
+    Integer defaultDiskSnapshotMaxMonthly = 12;
+    Boolean defaultDiskSnapshotUsageVirtualsizeSelect = false;
+    Integer defaultBackupRetry = 3;
+    Integer defaultBackupRetryInterval = 300;
+    Integer snapshotDeltaMax = 16;
+
+    ConfigKey<Integer> snapshotHourlyMax = new ConfigKey<Integer>(Integer.class,
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManager.java]: why did you remove the static and  final keywords here? 

It is rare to see a proper use of them in ACS, and here it was being used in the right manner.

File: server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java [hunk 9]
```diff
@@ -1045,16 +1037,13 @@ protected VMSnapshot retrieve() {
...
-            super(VirtualMachine.class, job, VmJobCheckInterval.value(), new Predicate() {
-                @Override
-                public boolean checkCondition() {
-                    AsyncJobVO jobVo = _entityMgr.findById(AsyncJobVO.class, job.getId());
-                    assert (jobVo != null);
-                    if (jobVo == null || jobVo.getStatus() != JobInfo.Status.IN_PROGRESS)
-                        return true;
+            super(VirtualMachine.class, job, VirtualMachineManager.VmJobCheckInterval.value(), () -> {
+                AsyncJobVO jobVo = _entityMgr.findById(AsyncJobVO.class, job.getId());
+                assert (jobVo != null);
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java]: Can you remove this `assert`?

File: server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java [hunk 10]
```diff
@@ -1045,16 +1037,13 @@ protected VMSnapshot retrieve() {
...
+                if (jobVo == null || jobVo.getStatus() != JobInfo.Status.IN_PROGRESS)
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/vm/snapshot/VMSnapshotManagerImpl.java]: instead of using a conditional like "if true", return true, otherwise, return false; what if we used `return jobVo == null || jobVo.getStatus() != JobInfo.Status.IN_PROGRESS`

**Comment by rafaelweingartner** [comment]
@bwsw are you willing to help us getting this PR into 4.12?

**Comment by GabrielBrascher** [comment]
Removing this PR from milestone 4.12.0.0.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManager.java (hunk 8 above)
Comment (DennisKonrad) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManager.java]: @rafaelweingartner Can you explain why the keywords are needed here? I would like to understand your comment

**Comment by rhtyd** [comment]
Can you rebase and fix the conflicts against latest master @bwsw

**Comment by DaanHoogland** [comment]
@bwsw still valueable work, this? can you rebase and add 4.15 as milestone if you think it is?

**Comment by DaanHoogland** [comment]
@bwsw can you rebase (in light of 4.15 merge)?

**Comment by bwsw** [comment]
@DaanHoogland Daan, as there is a bad-looking feature implementation architecture, I think we have to engineer it from scratch. I'll ask a colleague to take a look at this soon.

**Comment by DaanHoogland** [comment]
ok, tnx @bwsw , sounds like an FS page is in order. If you think that is overdone just force push ;)

**Comment by DaanHoogland** [comment]
@bwsw any updates?

**Comment by sureshanaparti** [comment]
ping @bwsw , any update on this?

**Comment by rohityadavcloud** [comment]
Hi @bwsw thank you for your PR, given the inactivity in the last 3+ years, I'm closing this. If you happen to work on this feature and can fix conflicts pl re-open or re-raise the PR.
